composer/models/ssd/ssd.py

```diff
@@ -0,0 +1,161 @@
+
+        self.loss_func = Loss(dboxes)
+        self.MAP = my_map()
+
+    def loss(self, outputs: Any, batch: BatchPair) -> Tensors:
+
+        (img, img_id, img_size, bbox, label) = batch
+        trans_bbox = bbox.transpose(1, 2).contiguous()
+
+        ploc, plabel = outputs
+        gloc, glabel = Variable(trans_bbox, requires_grad=False), \
+                        Variable(label, requires_grad=False)
+
+        loss = self.loss_func(ploc, plabel, gloc, glabel)
+        return loss
+
+    def metrics(self, train: bool = False) -> Metrics:
+        return self.MAP
+
+    def forward(self, batch: BatchPair) -> Tensor:
+        (img, img_id, img_size, bbox, label) = batch
+        context = contextlib.nullcontext if self.training else torch.no_grad
+
+        img = Variable(img, requires_grad=True)
+        ploc, plabel = self.module(img)
+
+        return ploc, plabel
+
+    def validate(self, batch: BatchPair) -> Tuple[Any, Any]:
+
+        dboxes = dboxes300_coco()
+        input_size = 300
+        val_trans = SSDTransformer(dboxes, (input_size, input_size), val=True)
+        data = "/localdisk/coco"
+
+        val_annotate = os.path.join(data, "annotations/instances_val2017.json")
+        val_coco_root = os.path.join(data, "val2017")
+        train_annotate = os.path.join(data, "annotations/instances_train2017.json")
+        train_coco_root = os.path.join(data, "train2017")
+        val_coco = COCODetection(val_coco_root, val_annotate, val_trans)
+
+        inv_map = {v: k for k, v in val_coco.label_map.items()}
+
+        ret = []
+
+        overlap_threshold = 0.50
+        nms_max_detections = 200
+
+        dboxes = dboxes300_coco()
+        encoder = Encoder(dboxes)
+
+        (img, img_id, img_size, _, _) = batch
+        with torch.no_grad():
+            ploc, plabel = self.module(img)
+            ploc, plabel = ploc.float(), plabel.float()
+
+            for idx in range(ploc.shape[0]):
+                ploc_i = ploc[idx, :, :].unsqueeze(0)
+                plabel_i = plabel[idx, :, :].unsqueeze(0)
+
+                try:
+                    result = encoder.decode_batch(ploc_i, plabel_i, 0.50, 200)[0]
+                except:
+                    # raise
+                    print("")
+                    print("No object detected in idx: {}".format(idx))
+                    continue
+
+                htot, wtot = img_size[0][idx].item(), img_size[1][idx].item()
+                loc, label, prob = [r.cpu().numpy() for r in result]
+                for loc_, label_, prob_ in zip(loc, label, prob):
+                    ret.append([int(img_id[idx].detach().cpu()), loc_[0] * wtot, \
+                                loc_[1] * htot,
+                                (loc_[2] - loc_[0]) * wtot,
+                                (loc_[3] - loc_[1]) * htot,
+                                prob_,
+                                inv_map[label_]])
+
+        return ret, ret
+
+
+class my_map(Metric):
```

Comment:
torchmetrics has a MAP metric already. may be better to use this to avoid depending on COCO since loading it is odd. https://torchmetrics.readthedocs.io/en/latest/references/modules.html#meanaverageprecision